### ISSUE TITLE
"Exception: No query supplied" in the listing block

### ISSUE THREAD

**Issue by wesleybl** [plone/volto#3839]
**Describe the bug**
When we load a page that has a listing block, without a search criteria, we get the error below in Plone:

```bash
2022-11-04 16:51:47 ERROR [Zope.SiteErrorLog:35][waitress-3] Exception: http://localhost:3000/@querystring-search
Traceback (innermost last):
  Module ZPublisher.WSGIPublisher, line 167, in transaction_pubevents
  Module ZPublisher.WSGIPublisher, line 376, in publish_module
  Module ZPublisher.WSGIPublisher, line 271, in publish
  Module ZPublisher.mapply, line 85, in mapply
  Module ZPublisher.WSGIPublisher, line 68, in call_object
  Module plone.rest.service, line 22, in __call__
  Module plone.restapi.services, line 19, in render
  Module plone.restapi.services.querystringsearch.get, line 30, in reply
Exception: No query supplied
```

Who makes this call is [getAsyncData.js](https://github.com/plone/volto/blob/master/src/components/manage/Blocks/Listing/getAsyncData.js)

It sends the `query` parameter as an empty list.

It looks like it's an attempt to render the block server-side, but I don't quite understand it. The fact is that the block works without getAsyncData. So I'm not quite sure what the idea is here.

**To Reproduce**

1. Create a listing block at home, without any search criteria.
2. Press F5 on home.
3. See the Plone log.

**Expected behavior**
No error in Plone.

**Software (please complete the following information):**

- OS: Ubuntu 22
- Browser: Chrome
- Volto Version: 16.0.0-alpha.48
- Plone Version 6.0.0b3
- Plone REST API Version: 8.30.0

**Comment by tiberiuichim** [comment]
@wesleybl It's related to the server-side rendering of the listing block, see https://6.dev-docs.plone.org/volto/blocks/ssr.html

It's probably a bug, that code needs to be brought more in line with the rest of the listing block data fetching preparations.

**Comment by wesleybl** [comment]
@tiberiuichim i'm trying to modify the listing block, to return more calalog metadata. But I saw that there are requests in [getAsyncData.js](https://github.com/plone/volto/blob/master/src/components/manage/Blocks/Listing/getAsyncData.js) and in [withQuerystringResults.jsx](https://github.com/plone/volto/blob/master/src/components/manage/Blocks/Listing/withQuerystringResults.jsx#L80). Which of these files should I make changes to? Or on both?

**Comment by tiberiuichim** [comment]
@wesleybl both. 

getAsyncData is only used on the server-side rendering path. It's meant to populate the Redux store that will be used by the listing block components. Here's a page that hopefully will explain this mechanism: https://training.plone.org/effective-volto/addons/asyncconnect.html

withQuerystringResults is used by the listing block to fetch the data.

In principle these two functions should produce exactly the same query (pagination excluded, as there's no URL-based set current page). If they don't, it's a bug in the getAsyncData and should be fixed.

**Comment by wesleybl** [comment]
@tiberiuichim as I understand it, this is a mechanism to prevent the page from loading without the listing. But on the Browser side, the listing can be modified, after server rendering. That's it?

Well it looks like the way it is today, when we don't have a criterion, nothing is rendered on the server side (`getAsyncData.js`), since the query parameter is empty [here](https://github.com/plone/volto/blob/63bbbc0ae3ae0b8a381c9f1f1d04a1e761340e77/src/components/manage/Blocks/Listing/getAsyncData.js#L12) (causing the described error).

The listing only appears because `withQuerystringResults.js` has a fallback:

https://github.com/plone/volto/blob/20840fdc5fc5a80e77e53ac04a028b613d444eaf/src/components/manage/Blocks/Listing/withQuerystringResults.jsx#L104-L106

when we don't have query.

It may be necessary to create a "default query", both in `withQuerystringResults.js` and in `getAsyncData.js`, when there is no query. Maybe in editing this "default query" should already be saved.

**Comment by wesleybl** [comment]
@tiberiuichim @sneridagh any opinions here? I propose the default query:

```javascript
const defaultQuery = [
  {
    i: 'path',
    o: 'plone.app.querystring.operation.string.relativePath',
    v: '::1',
  },
];
```

when we have no defined criteria. This query will have the same result as the items key of the content.

**Comment by tiberiuichim** [comment]
@wesleybl Ok, +1

**Comment by mliebischer** [comment]
Hello @tiberiuichim and @davisagli, we have noticed that the [current solution in Volto 18](https://github.com/plone/volto/commit/3acdf4dd7eeb6e63b1b29f0d82cb304402058310) does not work in every situation.

### Steps to reproduce:

1. add listing block
2. add a location criteria
3. then set the depth to 1 
![Image](https://github.com/user-attachments/assets/d06f982a-8176-4cf9-9798-b4b68e2b67ba)
5. delete the location criteria again (the depth criteria is now invisible, but remains set)

I then had a look at the `data`  variable in [getAsyncData.js](https://github.com/plone/volto/blob/f78ce941cd929bedb82cf1a8cf7b5da830dea207/packages/volto/src/components/manage/Blocks/Listing/getAsyncData.js#L20):
```javascript
{
  '@type': 'listing',
  headlineTag: 'h2',
  querystring: { depth: '1', query: [], sort_order: 'ascending' },
  styles: {},
  variation: 'default'
}
```

Problem: The backend raises an exception if `query` is empty: 
https://github.com/plone/plone.restapi/blob/bc29368009dfc0719adab723cfddd9675a3c79ea/src/plone/restapi/services/querystringsearch/get.py#L34-L44

But the current solution in Volto only checks if the whole `querystring` variable is empty, not if the `query` in it is empty:
https://github.com/plone/volto/blob/f78ce941cd929bedb82cf1a8cf7b5da830dea207/packages/volto/src/components/manage/Blocks/Listing/getAsyncData.js#L27

A better solution could be:
```javascript
if (!data.querystring || data.querystring.query.length === 0) {
```

Perhaps the issue should be reopened (or a new issue created).  🙏

**Comment by stevepiercy** [comment]
@mliebischer would you please test the pull request https://github.com/plone/volto/pull/3866#issuecomment-2751704208 to see if it resolves your report, and reply in that pull request? Thank you!

**Comment by wesleybl** [comment]
@mliebischer I think I've seen the problem you described in [#90](https://github.com/plone/volto/issues/3839#issuecomment-2637410384) before, even before fix [#4726](https://github.com/plone/volto/pull/6423). See:

https://github.com/plone/volto/pull/3866#issuecomment-1359318005
